Add project skeleton

i-cod opened the pull request:
**About service**
The purpose of the `github-status` service is to send messages about incidents in _Github_ components to _VKTeams_.
The service does not have a UI. It is assumed that the service will be launched in the infrastructure and will run as a background process inside the docker container.

**The logic of the service**
The service is based on an endless loop in which the following actions are performed:
- With a specified interval (by default 30 seconds), an api request is executed to check the status of Github components
- If an incident occurs in any component, information about it is added to the internal incident registry
- If the incident is relevant for more than 2 minutes, a fail message is sent to the VKTeams channel
- If all incidents are fixed, a message is sent that everything is fixed

Closes [#166](https://github.com/tarantool/infra/issues/166)

Commit:
Add project skeleton

About service
The purpose of the github-status service is to send messages about
incidents in Github components to VKTeams.
The service does not have a UI. It is assumed that the service
will be launched in the infrastructure and will run as a
background process inside the docker container.

The logic of the service
The service is based on an endless loop in which the following
actions are performed:
- With a specified interval (by default 30 seconds), an api
request is executed to check the status of Github components
- If an incident occurs in any component, information about it
is added to the internal incident registry
- If the incident is relevant for more than 2 minutes, a fail
message is sent to the VKTeams channel
- If all incidents are fixed, a message is sent that everything
is fixed

Part of #166

ochaplashkin commented on settings.py:
```diff
@@ -0,0 +1,24 @@
+import os
+
+
+class Config:
```
I like `dataclass`:

```python
from dataclasses import dataclass

@dataclass
class SubConfig:
    bar: str

@dataclass
class Config:
    foo: str
    internal: SubConfig
    
def read_config(path: str) -> Config:
    # do something from config file
    # or
    return Config(
            foo=os.environ.get("FOO")
            internal=SubConfig(bar=os.environ.get("BAR"))
          )
```

i-cod replied:
Fixed

ochaplashkin commented on watcher/client.py:
```diff
@@ -0,0 +1,22 @@
+import requests
+
+
+class Client:
```
May be `Client` -> `GitHubClient` or `GitHubAPI`?
`Client` is too general a concept

i-cod replied:
Fixed to `GithubClient`

ochaplashkin commented on watcher/incident_register.py:
```diff
@@ -0,0 +1,34 @@
+
+
+@dataclass
+class Incident:
```
I would add a **UID** field (it would be the key for the dictionary)

i-cod replied:
It will be useless in this logic context

ochaplashkin commented on watcher/watcher.py:
```diff
@@ -0,0 +1,59 @@
+        while True:
+            func(*args, **kwargs)
+            _self = args[0]
+            time.sleep(_self.config.status_checking_interval)
```
implicit behavior

i-cod replied:
Fixed

Commit:
Add dockerfile

Added dockerfile based on python:3.10-slim image. In configuration
installing dependencies and running the service.

Image build
docker build -f ./Dockerfile -t githubstatus .

Running the container
docker run --env BOT_TOKEN=<token> --env CHAT_ID=<chat_id> githubstatus

Part of #166

ylobankov commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+COPY . .
+RUN pip3 install -r requirements.txt
+
+CMD ["python3", "main.py"]
```
Please add an empty line after this.

i-cod replied:
Fixed

ylobankov commented on watcher/incident_registry.py:
```diff
@@ -0,0 +1,41 @@
+                return incident
+        return None
+
+    def change_notification(self, component_name: str, value: bool):
```
Shouldn't it be named `update`?

i-cod replied:
Fixed

ylobankov replied:
I meant just `update`.

ylobankov replied:
Also, it would be good to support updating not only one particular field, but updating a few fields at once.

i-cod replied:
Fixed

Commit:
Add flake8 and unit tests

Tests are based on PyTest.

Run test
pip install -r test-requirements.txt
python -m pytest -v test/

Closes #166

ylobankov commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+mailru-im-bot==0.0.21
+requests==2.26.0
+retry==0.9.2
+requests-mock==1.10.0
```
Please separate test requirements from main requirements. Usually, projects have `requirements.txt` and `test-requirements.txt`.

i-cod replied:
Fixed

ochaplashkin commented on test-requirements.txt:
```diff
@@ -0,0 +1,2 @@
+requests-mock==1.10.0
```
Miss 
```
mailru-im-bot
requests
retry
```

i-cod replied:
It's all contained in the `requirements.txt`

i-cod replied:
Fixed

ylobankov commented on README.md:
```diff
@@ -1 +1,17 @@
\ No newline at end of file
+# github-status
+
+## Deployment
```
Let's include deployment part to the previous commit.

i-cod replied:
Fixed

ylobankov commented on README.md:
```diff
@@ -1 +1,17 @@
+
+`docker run --env BOT_TOKEN=<token> --env CHAT_ID=<chat_id> githubstatus`
+
+## Run flake8
```
Let's include testing part to the next commit. So this commit should disappear.

i-cod replied:
Fixed

ochaplashkin commented on README.md:
```diff
@@ -1 +1,17 @@
+`docker run --env BOT_TOKEN=<token> --env CHAT_ID=<chat_id> githubstatus`
+
+## Run flake8
+`flake8 --extend-ignore E501 --exclude venv/`
```
It seems to me that ignoring the rules is not the best approach:
```
$ flake8 --exclude venv/
./main.py:6:80: E501 line too long (87 > 79 characters)
./settings.py:36:80: E501 line too long (82 > 79 characters)
./settings.py:37:80: E501 line too long (93 > 79 characters)
./settings.py:38:80: E501 line too long (82 > 79 characters)
./settings.py:39:80: E501 line too long (102 > 79 characters)
./settings.py:40:80: E501 line too long (101 > 79 characters)
./test/test_client.py:17:80: E501 line too long (113 > 79 characters)
```

Let's create a [config for flake8](https://flake8.pycqa.org/en/2.5.5/config.html#user-global) and add the string length there (this will already be a rule for the project) and list of excluding directories.

i-cod replied:
Config file added

ochaplashkin commented on test/conftest.py:
Why do we need to add an empty file?

Ideally, fixtures for tests should be stored here (I know that it won't work without it, because py test searches the directory tree)

i-cod replied:
Removed

ochaplashkin commented on README.md:
```diff
@@ -1 +1,17 @@
+
+## Run tests:
+`pip install -r test-requirements.txt`\
+`python -m pytest -v test/`
```
It looks strange, why is `CHAT_ID` required from me when testing locally?

```
$ pip install -r test-requirements.txt
...
$ python -m pytest -v test/
============================= test session starts ==============================
platform linux -- Python 3.9.1, pytest-7.3.1, pluggy-1.2.0 -- /home/pc/workspace/vk/github-status/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/pc/workspace/vk/github-status
plugins: requests-mock-1.10.0
collected 5 items / 2 errors                                                   

==================================== ERRORS ====================================
...
=========================== short test summary info ============================
ERROR test/test_client.py - AssertionError: Environment variable CHAT_ID not set.
ERROR test/test_vkteamsbot.py - AssertionError: Environment variable CHAT_ID not set.
!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 2 errors in 0.10s ===============================
```

i-cod replied:
Fixed

ochaplashkin reviewed (approved):
LGTM :rocket:

ylobankov commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+mailru-im-bot==0.0.21
+requests==2.26.0
+retry==0.9.2
+flake8==6.0.0
```
It is a dependency for testing. Looks like it shouldn't be here.

i-cod replied:
Fixed

ylobankov replied:
You should remove that in this commit, not in the second one.

ylobankov commented on README.md:
```diff
@@ -1 +1,10 @@
+# github-status
+
+## Deployment
+Build Docker container with:
```
container -> image

i-cod replied:
Fixed

ylobankov commented on test/test_client.py:
```diff
@@ -0,0 +1,79 @@
+            except BadRequestException:
+                return
+            else:
+                assert True
```
Looks strange. As far as I understand, we should fail the test here.

i-cod replied:
Removed in all places

ylobankov commented on test/test_client.py:
```diff
@@ -0,0 +1,79 @@
+            except BadRequestException:
+                return
+            else:
+                assert True
```
Looks strange. As far as I understand, we should fail the test here.

ylobankov commented on test/test_vkteamsbot.py:
```diff
@@ -0,0 +1,53 @@
+            except BadRequestException:
+                return
+            else:
+                assert True
```
Looks strange. As far as I understand, we should fail the test here.

ylobankov commented:
Please provide a description for the third commit like you do for the second one.

ylobankov commented:
Please fix the first commit message:

```
The logic of the service
- The service is based on an endless loop in which the following
actions are performed:
- With a specified interval (by default 30 seconds), an api
request is executed to check the status of Github components
...
```
->
```
The logic of the service
The service is based on an endless loop in which the following
actions are performed:
- With a specified interval (by default 30 seconds), an api
request is executed to check the status of Github components
...
```

Please fix the third commit message to be consistent with the second one:

```
Add flake8 and unit tests

Tests are based on PyTest.

Run test
pip install -r test-requirements.txt
python -m pytest -v test/
